daemon/monitor.go

```diff
@@ -85,16 +85,29 @@ func (daemon *Daemon) StateChanged(id string, e libcontainerd.StateInfo) error {
 		}
 		return daemon.postRunProcessing(c, e)
 	case libcontainerd.StateExitProcess:
+		closeAsync := make(chan bool)
+		waitClose := false
 		c.Lock()
-		defer c.Unlock()
-		if execConfig := c.ExecCommands.Get(e.ProcessID); execConfig != nil {
-			ec := int(e.ExitCode)
-			execConfig.ExitCode = &ec
-			execConfig.Running = false
-			execConfig.Wait()
-			if err := execConfig.CloseStreams(); err != nil {
-				logrus.Errorf("%s: %s", c.ID, err)
+		defer func() {
+			c.Unlock()
+			if waitClose {
+				<-closeAsync
```

Comment:
See my other comment below first, as this concern is invalidated if we don't need to have the container lock here at all.

I'd also prefer not to have this synchronization. I added it in order to not change the behaviour of this function (return after Streams are closed), as I did not know the implications on Linux. If you think that it's ok to remove that works for me, as the process has already exited on Windows, and this is just waiting for the pipes to flush.